src/core/lib/slice/slice_internal.h

```diff
@@ -23,17 +23,168 @@
+  }
+  SliceRefcount(grpc_slice_refcount* sub, Type type)
+      : sub_refcount_(sub), ref_type_(type) {
+    // Initializer for static.
+  }
+
+  Type getType() const { return ref_type_; }
+
+  int Eq(const grpc_slice& a, const grpc_slice& b);
+
+  uint32_t Hash(const grpc_slice& slice);
+  void Ref() {
+    if (!ref_) return;
+#ifndef NDEBUG
+    if (ensureRefNonZero_) {
+      ref_->RefNonZero();
+      return;
+    }
+#endif
+    ref_->Ref();
+  }
+  void Unref() {
+    if (!ref_) return;
+    if (ref_->Unref()) {
+      dest_fn_(destroy_fn_arg_);
+    }
+  }
+
+  bool Validate() const { return ref_ != nullptr; }
+
+ private:
+  friend struct ::grpc_slice_refcount;
+
+  grpc_slice_refcount* sub_refcount_ = nullptr;
+  Type ref_type_ = REGULAR;
+  RefCount* ref_ = nullptr;
+#ifndef NDEBUG
+  bool ensureRefNonZero_ = false;
+#endif
+  DestroyerFn dest_fn_ = nullptr;
+  void* destroy_fn_arg_ = nullptr;
+};
+
+}  // namespace grpc_core
+
+struct grpc_slice_refcount {
+  grpc_slice_refcount() { impl_.sub_refcount_ = this; }
+  grpc_slice_refcount(grpc_core::SliceRefcount::Type type,
+                      grpc_core::RefCount* ref, bool ensureRefNonZero,
+                      grpc_core::SliceRefcount::DestroyerFn destroyer_fn,
+                      void* destroyer_arg, grpc_slice_refcount* sub)
+      : impl_(type, ref, ensureRefNonZero, destroyer_fn, destroyer_arg, sub) {
+    // Nothing to do here.
+  }
+  grpc_slice_refcount(grpc_slice_refcount* sub,
+                      grpc_core::SliceRefcount::Type type)
+      : impl_(sub, type) {
+    // Nothing to do here.
+  }
+
+  grpc_slice_refcount* sub_refcount() const { return impl_.sub_refcount_; }
+
+  grpc_core::SliceRefcount impl_;
+};
+
+namespace grpc_core {
+
+struct InternedSliceRefcount {
+  static void Destroy(void* arg) {
+    static_cast<InternedSliceRefcount*>(arg)->DestroyInstance();
+  }
+
+  InternedSliceRefcount(size_t length, uint32_t hash,
+                        InternedSliceRefcount* bucket_next)
+      : base(grpc_core::SliceRefcount::Type::INTERNED, &refcnt,
+             /*ensureRefNonZero=*/false, Destroy, this, &sub),
+        sub(grpc_core::SliceRefcount::Type::REGULAR, &refcnt,
+            /*ensureRefNonZero=*/false, Destroy, this, &sub),
+        length(length),
+        hash(hash),
+        bucket_next(bucket_next) {
+    // Nothing else to do here.
+  }
+
+  void DestroyInstance();
+
+  grpc_slice_refcount base;
+  grpc_slice_refcount sub;
+  size_t length;
+  RefCount refcnt;
+  uint32_t hash;
+  InternedSliceRefcount* bucket_next;
+};
+
+inline int SliceRefcount::Eq(const grpc_slice& a, const grpc_slice& b) {
+  switch (ref_type_) {
+    case STATIC:
+      return GRPC_STATIC_METADATA_INDEX(a) == GRPC_STATIC_METADATA_INDEX(b);
+    case INTERNED:
+      return a.refcount == b.refcount;
+    case REGULAR:
+      break;
+  }
+  if (GRPC_SLICE_LENGTH(a) != GRPC_SLICE_LENGTH(b)) return false;
+  if (GRPC_SLICE_LENGTH(a) == 0) return true;
```

Comment:
for my education: why is this correct? can you add a comment please?

Reply:
Actually there's two opportunities here.

Regarding the actual explanation, if two slices are of different lengths, they must be unequal (hence false). Else: same length. If that length is 0, trivially they are equal, so return true. Otherwise do the memcmp.

Opportunities:
#1: We can remove the len() == 0 check, since when 0 is passed into memcmp, we by default get 0 back.

#2: grpc_slice_eq by default is what people externally would use; that only uses this new Eq() method if the types are the same (in the old days: if the vtables were identical).
So if we compare, say, an interned slice to a regular slice, we'll call grpc_slice_default_eq_impl(grpc_slice a, grpc_slice b).

That is rather costly; it has the old style copy values invocation, and just does what we do in the fallthrough case of Eq() anyways.

So instead, I propose just using the new Eq() method always since it will always work even if the types are different.
ie. Suppose slice::a is static, but b is not. GRPC_STATIC_METADATA_INDEX(b) is just some pointer arithmetic followed by  a compare; the pointer itself is never deref'd. So that case is safe.
ie. Suppose slice::a is interned, but b is not. In general the only way two slices share the same refcount is if they're stream_refs or if they're interned. It should not be possible for a to be interned and for b to share the ref if b is not interned.

Then grpc_slice_default_eq_impl() still exists (being a part of the public API) but we can be faster by not using it. Thoughts?

Reply:
Actually, never mind this comment. I forgot to resolve this. Thanks